woocommerce/woocommerce-gutenberg-products-block
# Convert apple pay integration to payment request integration and finish implementation

Over the weekend I did an experiment to see how hard it would be to just implement the Stripe payment request api completely instead of just apple pay. Turns out, it wasn't hard at all (just needed to tweak things a bit).

In this pull:

### The full stripe PaymentRequest api is supported (for express payments)

This means we have full support for ApplePay, GooglePay, Microsoft pay and any payment method that supports the payment request api (as made available via stripe).  In particular the utility of having google pay support means for easier local development/testing as long as you have https:// enabled.

### Complete checkout flow is supported for express payments

Here's a list of expectations that are supported by the work in this pull.

- All _successful_ payments are tested and work. I have not verified failures or error handling in payment methods yet because that increases the scope of testing for this pull, followups will need to test and make any necessary fixes. Error handling and fail logic _is_ in place, it's just not tested.
- Shipping is supported for express payment methods (when shipping is required for the order). So selecting an address should result in rates getting updated in the payment request interface and selecting rates should update the totals in the payment request interface.
- Making a payment using the payment request interface should complete successfully and update the order with the correct details (except for the known issue as outlined in #2139)
- Shipping package support is _not_ available with this work (see followups section).
- Server side processing utilizes the stripe payment gateway (request payment method) for handling the extra details for express payments passed through via checkout.
- **Note:** When express payment methods are being used to make a payment, checkout field validation is disabled. This is because the express payment method has its own internal validation and things break with the current order of processing that includes the checkout form validation (the form is typically _empty_ and thus fails validation before payment processing is triggered for express payment method).

~**Known issue with ApplePay and purchases involving shipping**~ (fixed by #2163)

It looks like applePay doesn't do any validation of state/province or postal code, so it's possible for user's to type whatever they want in the field (at least applePay in the browser). Since our shipping address hook (useShippingAddress) uses a comparison of the previous address to the new address (specific address fields) to determine whether to update shipping rates, this means on successful payment, the passed through shipping field will trigger a new shipping rate request and this causes checkout state to go back to idle (which eventually triggers a bunch of shipping rate and checkout processing requests as states flip around).  

So after more testing, I discovered that the issue here is the `shippingaddresschange` event in stripe may not contain all address details - and even could contain partial/redacted address details (see [here](https://stripe.com/docs/js/payment_request/events/on_shipping_address_change)). I was using that to get shipping rates (and it will only work partially - rates specific to zip code won't be guaranteed to work). It does look like the address provided on the `source` event (via the source object) _is_ correct except ApplePay provides the full state/province name, not the state code. Another issue is that when the shipping request happens after payment is successful, the checkout calculating state is incremented which changes checkout status to calculating. After calculating is decremented, the status goes back to the previous state. This is clever and allows for less logic in the code for setting state, but the problem here is that the previous state becomes `checkoutProcessingComplete` and this causes the `processCheckout` action in the `CheckoutProcessor` component to fire again (which in turn will trigger the payment success event in the payment method again which in turn causes shipping rates to recalculate etc (creating a bit of a loop). So I may have to separate calculating state from the actual checkout processing status so that it's possible for checkout to be processing and calculating or not calculating (similar to how checkout has has an error state separate from it's status). So essentially ending up with three checkout tracked states:

- Error state: checkout either has an error (derived from other error states) or not.
- Calculating state: derived from all loading states that contribute to total changes (`isCalculating` would simply be whether there is a positive value for the calculating increments).
- Flow state: which is one of `PRISTINE`, `IDLE`, `PROCESSING`, `PROCESSING_COMPLETE`, `COMPLETE`. At any point in the checkout flow the checkout could be also calculating or have an error. I think that will make things a bit more clear and precise (and prevent some of the loops I'm seeing).

## To Test

> Note: you can use the ephemeral site I setup for testing if you want: https://ephemeral-nerrad-20200302.atomicsites.blog/

For the express payment methods, probably the easiest to test is using Google Pay (chrome browser) as it can be tested locally as long as you have ssl enabled for your local dev site.  I'm fine with merging if Google Pay works okay (but I am going to test Apple Pay today as well).

- See the list of expectations and verify that it works as expected for making payments with express payment methods. Keep the console open while testing and watch for errors. Also watch that there aren't more network requests than expected.
- Verify that cc and cheque payment methods still process payments successfully and that validation works as expected for non express payments.

## Followups

- stripe.paymentRequest payments do not handle multiple packages. This will need to be handled in a followup. We'll either have to consider disabling the payment method if there are multiple packages in the order/cart, or setup the [shipping options](https://stripe.com/docs/js/appendix/shipping_option) for the payment method so each option is a combination of the various package selections (since only one option can be selected). The latter can be super complex and grow exponentially depending on the number of packages might exist.  I'll create an issue as a followup after this is merged.  I should note that I did look at the existing stripe payment gateway extension and how they handle packages is a bit puzzling too...so this may be something that's already a bit of a known issue for express payment methods (might need a p2).
- the `canPay` config option currently is using hardcoded country and currency code values, this will need updated to use something provided by the server so it's more context aware of the shopper in the checkout.
- currently express payment methods need to take care of setting active payment method when their payment request button is clicked and also restore active payment method when the payment request modal is closed. In this pull the payment request express component is handling this logic but ideally it would be moved somehow to checkout so that payment methods don't have to manage (or forget!) that complexity.
- as discussed in [this comment thread](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/2127#discussion_r404891414) the current implementation of deriving server side payment_method id needs reworked. This will involve ensuring client side is always passing along the correct payment_method for executing payment processing (instead of leaving it up to payment methods to correctly define as a part of the `paymentData` package in _some_ scenarios - namely stripe express). There's potentially some overlap with this item and the previous one.
